### PR TITLE
Feat(eos_designs): Add support to add access-groups on l3-interfaces through network-services

### LINKED ISSUE #3534: Add the ability to add access-group membership in network services
### Enhancement summary

It would be nice when defining a L3 / SVI in network services to also be able to add an access group right there. That was the tags take care of only adding the L3/SVI as it already does but also only add the access group to the devices that need it. 

### Which component of AVD is impacted

eos_designs

### Use case example

It would be nice when defining a L3 / SVI in network services to also be able to add an access group right there. That was the tags take care of only adding the L3/SVI as it already does but also only add the access group to the devices that need it. in-network

### Describe the solution you would like

It would be nice when defining a L3 / SVI in network services to also be able to add an access group right there. That was the tags take care of only adding the L3/SVI as it already does but also only add the access group to the devices that need it. in-network

### Describe alternatives you have considered

_No response_

### Additional context

_No response_

### Contributing Guide

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by laxmikantchintakindi**
## Change Summary
Add support to add access-groups on l3-interfaces through network-services

## Related Issue(s)

Fixes #3534
Only address the L3 interfaec but not the SVIs as it was already implemented in https://github.com/aristanetworks/avd/pull/4096.

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes

Initial ask
> It would be nice when defining a L3 / SVI in network services to also be able to add an access group right there. That was the tags take care of only adding the L3/SVI as it already does but also only add the access group to the devices that need it.

This PR covers the L3 interfaces part in network services - SVIs was done in https://github.com/aristanetworks/avd/pull/4096

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 1]
```diff
@@ -682,6 +682,50 @@ def _svi_acls(self: AvdStructuredConfigNetworkServices) -> dict[str, dict[str, d
...
+            return None
+
+        l3_interface_acls = {}
+        for tenant in self.shared_utils.filtered_tenants:
+            for vrf in tenant["vrfs"]:
+                for l3_interface in vrf["l3_interfaces"]:
+                    for interface in l3_interface["interfaces"]:
+                        ipv4_acl_in = get(l3_interface, "ipv4_acl_in")
+                        ipv4_acl_out = get(l3_interface, "ipv4_acl_out")
+                        if ipv4_acl_in is None and ipv4_acl_out is None:
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py:716]: In general why we need to use both in and out ACL?
@ClausHolbechArista

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 2]
```diff
@@ -682,6 +682,50 @@ def _svi_acls(self: AvdStructuredConfigNetworkServices) -> dict[str, dict[str, d
...
+                            continue
+                        interface_name = interface
+                        interface_idx = l3_interface["interfaces"].index(interface_name)
+                        interface_ip: str | None = l3_interface["ip_addresses"][interface_idx]
+                        if interface_ip is not None and "/" in interface_ip:
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: i know you are following the above svi logic but we can have better method to get the ip from prefix.
```
if interface_ip is None:
  continue
ip_address = str(ip_interface(interface_ip).ip)

Comment (MaheshGSLAB) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: from ipaddress import ip_interface

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Fixed.

**Review by MaheshGSLAB — DISMISSED** [review]

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 3]
```diff
@@ -688,6 +688,14 @@ $defs:
                       type: int
                       convert_types:
                         - str
+                    ipv4_acl_in:
+                      type: str
+                      # convert_type:
+                      #   - int
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: remove comments

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py (hunk 1 above)
Comment (gmuloc) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py:716]: yes

**Review by gmuloc — CHANGES_REQUESTED** [review]
So the PR only partially fix the issue as it is asking for l3_interfaces and SVIs and the PR currently implements only l3_interfaces.

I feel like the code to compute the access_list is a bit of a repeat from the one we have for l3_interfaces under node but maybe there was no easy way to reuse it?

Otherwise it looks good to me if we accept the fact that it is a partial fix.

Can you please remove the comments from the schema?

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py (hunk 1 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py:716]: This is the check to `continue` so we only skip the rest of the code if both in and out are `None`.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 4]
```diff
@@ -682,6 +682,51 @@ def _svi_acls(self: AvdStructuredConfigNetworkServices) -> dict[str, dict[str, d
...
+                    for interface in l3_interface["interfaces"]:
+                        ipv4_acl_in = get(l3_interface, "ipv4_acl_in")
+                        ipv4_acl_out = get(l3_interface, "ipv4_acl_out")
+                        if ipv4_acl_in is None and ipv4_acl_out is None:
+                            continue
+                        interface_name = interface
+                        interface_idx = l3_interface["interfaces"].index(interface_name)
+                        interface_ip: str | None = l3_interface["ip_addresses"][interface_idx]
+                        if interface_ip is None:
+                            continue
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Why are we continuing here? Not all ACLs use substitution.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 5]
```diff
@@ -682,6 +682,51 @@ def _svi_acls(self: AvdStructuredConfigNetworkServices) -> dict[str, dict[str, d
...
+        for tenant in self.shared_utils.filtered_tenants:
+            for vrf in tenant["vrfs"]:
+                for l3_interface in vrf["l3_interfaces"]:
+                    for interface in l3_interface["interfaces"]:
+                        ipv4_acl_in = get(l3_interface, "ipv4_acl_in")
+                        ipv4_acl_out = get(l3_interface, "ipv4_acl_out")
+                        if ipv4_acl_in is None and ipv4_acl_out is None:
+                            continue
+                        interface_name = interface
+                        interface_idx = l3_interface["interfaces"].index(interface_name)
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Something seems wrong to me here. We are not checking if we are the node for this interface, so if you have multiple interfaces in the list, you would render ACLs for all of them.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_A.yml [hunk 6]
```diff
@@ -242,10 +242,14 @@ tenant_a:
             mtu: 9000
             enabled: True
             description: "test"
+            ipv4_acl_in: TEST-IPV4-ACL-WITH-IP-FIELDS-IN
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_A.yml]: the test output is very hard to follow, since the same IPs are being reused everywhere. Please make sure the IPs are unique for the l3_interfaces we use for this test, so we can validate the output.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/DC1-BL1B.yml [hunk 7]
```diff
@@ -531,6 +537,49 @@ vlans:
...
+  - action: permit
+    protocol: ip
+    source: 10.10.20.20
+    destination: any
+- name: TEST-IPV4-ACL-WITH-IP-FIELDS-OUT_Ethernet10
+  entries:
+  - remark: Some remark will not require source and destination fields.
+  - action: permit
+    protocol: ip
+    source: 10.10.30.10
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/DC1-BL1B.yml]: This seems to be the wrong IP for DC1-BL1B on Ethernet10, so I think this is proving my point from the other comment, that we should evaluate the `nodes` key and skip interfaces that are not for us.

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml (hunk 3 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: Removed the comment marks.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py (hunk 4 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Removed this condition.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py (hunk 5 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Fixed. Now it will render the acls that belong to those nodes only.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_A.yml (hunk 6 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_A.yml]: Added new interfaces for test.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/DC1-BL1B.yml (hunk 7 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/structured_configs/DC1-BL1B.yml]: Fixed this problem. Now it is rendering the acl for interfaces of DC1-BL1B  only.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/ethernet_interfaces.py [hunk 8]
```diff
@@ -75,6 +75,14 @@ def ethernet_interfaces(self: AvdStructuredConfigNetworkServices) -> list | None
                                 "flow_tracker": self.shared_utils.get_flow_tracker(l3_interface, "l3_interfaces"),
                             }
 
+                            if get(self._l3_interface_acls, self.shared_utils.hostname):
+                                interface.update(
+                                    {
+                                        "access_group_in": get(self._l3_interface_acls[self.shared_utils.hostname], f"{interface_name}.ipv4_acl_in.name"),
+                                        "access_group_out": get(self._l3_interface_acls[self.shared_utils.hostname], f"{interface_name}.ipv4_acl_out.name"),
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/ethernet_interfaces.py]: Preventing issue with `.` in the interface name. Make sure this is also done anywhere else where you use `get()` with the interface names (for code changed in this PR)
```suggestion
                                        "access_group_in": get(self._l3_interface_acls[self.shared_utils.hostname], f"{interface_name}..ipv4_acl_in..name", separator=".."),
                                        "access_group_out": get(self._l3_interface_acls[self.shared_utils.hostname], f"{interface_name}..ipv4_acl_out..name", separator=".."),
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 9]
```diff
@@ -686,6 +686,52 @@ def get_internet_exit_nat_pool_and_profile(
     def _filtered_internet_exit_policy_types(self: AvdStructuredConfigNetworkServices) -> list:
         return sorted(set(internet_exit_policy["type"] for internet_exit_policy in self._filtered_internet_exit_policies))
 
+    @cached_property
+    def _l3_interface_acls(self: AvdStructuredConfigNetworkServices):
+        """
+        Returns a dict of
+            <hostname> : {
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Since this code runs per device, we should not index acls for interfaces that do not belong to this device.

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py (hunk 9 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Fixed.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/ethernet_interfaces.py (hunk 8 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/ethernet_interfaces.py]: Fixed.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 10]
```diff
@@ -686,6 +686,51 @@ def get_internet_exit_nat_pool_and_profile(
     def _filtered_internet_exit_policy_types(self: AvdStructuredConfigNetworkServices) -> list:
         return sorted(set(internet_exit_policy["type"] for internet_exit_policy in self._filtered_internet_exit_policies))
 
+    @cached_property
+    def _l3_interface_acls(self: AvdStructuredConfigNetworkServices):
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]:
```suggestion
    def _l3_interface_acls(self: AvdStructuredConfigNetworkServices) -> dict | None:
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 11]
```diff
@@ -686,6 +686,51 @@ def get_internet_exit_nat_pool_and_profile(
...
+
+        if not self.shared_utils.network_services_l3:
+            return None
+
+        l3_interface_acls = {}
+        for tenant in self.shared_utils.filtered_tenants:
+            for vrf in tenant["vrfs"]:
+                for l3_interface in vrf["l3_interfaces"]:
+                    for interface_idx, interface in enumerate(l3_interface["interfaces"]):
+                        ipv4_acl_in = get(l3_interface, "ipv4_acl_in")
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py:714]:
```suggestion
                        if l3_interface["nodes"][interface_idx] != self.shared_utils.hostname:
                            continue
                            
                        ipv4_acl_in = get(l3_interface, "ipv4_acl_in")
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 12]
```diff
@@ -686,6 +686,51 @@ def get_internet_exit_nat_pool_and_profile(
...
+                                    name=ipv4_acl_in,
+                                    interface_name=interface_name,
+                                    interface_ip=interface_ip,
+                                )
+                            if ipv4_acl_out is not None:
+                                l3_interface_acls.setdefault(interface_name, {})["ipv4_acl_out"] = self.shared_utils.get_ipv4_acl(
+                                    name=ipv4_acl_out,
+                                    interface_name=interface_name,
+                                    interface_ip=interface_ip,
+                                )
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]:
```suggestion
                        if ipv4_acl_in is not None:
                            l3_interface_acls.setdefault(interface_name, {})["ipv4_acl_in"] = self.shared_utils.get_ipv4_acl(
                                name=ipv4_acl_in,
                                interface_name=interface_name,
                                interface_ip=interface_ip,
                            )
                        if ipv4_acl_out is not None:
                            l3_interface_acls.setdefault(interface_name, {})["ipv4_acl_out"] = self.shared_utils.get_ipv4_acl(
                                name=ipv4_acl_out,
                                interface_name=interface_name,
                                interface_ip=interface_ip,
                            )
```

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 13]
```diff
@@ -691,6 +691,14 @@ $defs:
                       type: int
                       convert_types:
                         - str
+                    ipv4_acl_in:
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:694]: what about SVIs as requested in the initial issue?

Comment (gmuloc) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:694]: we will address SVI in another PR

File: python-avd/pyavd/_eos_designs/structured_config/network_services/ip_access_lists.py [hunk 14]
```diff
@@ -111,6 +111,12 @@ def ip_access_lists(self: AvdStructuredConfigNetworkServices) -> list | None:
                 for acl in interface_acls.values():
                     append_if_not_duplicate(ip_access_lists, "name", acl, context="IPv4 Access lists for SVI", context_keys=["name"])
 
+        if self._l3_interface_acls:
+            l3_interface_acls = self._l3_interface_acls
+            for l3_interface_acl in l3_interface_acls.values():
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/ip_access_lists.py]:
```suggestion
        if self._l3_interface_acls:
            for l3_interface_acl in self._l3_interface_acls.values():
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml (hunk 13 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:694]: https://github.com/aristanetworks/avd/blob/0b1272b9c4513aa99bdfbd986f5d5a69005fb267/python-avd/pyavd/_eos_designs/structured_config/network_services/ip_access_lists.py#L109

I think this already present.

Comment (gmuloc) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:694]: yes ! I just realized this. Thanks for commenting .. It was implemented in this PR: https://github.com/aristanetworks/avd/pull/4096

**Review by gmuloc — APPROVED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_A.yml [hunk 15]
```diff
@@ -255,6 +255,14 @@ tenant_a:
             enabled: True
             description: "Single description"
             descriptions: ["DC1-BL1A descriptions preferred over single description", "DC1-BL1B descriptions preferred over single description"]
+          - interfaces: [Ethernet12, Ethernet13, Ethernet12, Ethernet13]
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_A.yml]: Maybe we can add a test with subinterface to make sure we never break it here
```suggestion
          - interfaces: [Ethernet12, Ethernet13.10, Ethernet12, Ethernet13.10]
```

Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS/Tenant_A.yml]: Added.

**Review by MaheshGSLAB — APPROVED** [review]

**Review by ClausHolbechArista — APPROVED** [review]
